src/Raven.Server/Documents/Replication/ReplicationDocumentSender.cs

```diff
@@ -200,6 +202,59 @@ internal static IEnumerable<ReplicationBatchItem> GetReplicationItems(DocumentDa
             }
         }
 
+        private class ReplicationBatchState : PulsedEnumerationState<ReplicationBatchItem>
+        {
+            private ReplicationDocumentSender _parent;
+            private ReplicationState _replicationState;
+            public bool WasInterrupted { get; private set; }
+            public long Next { get; private set; }
+
+            public ReplicationBatchState(DocumentsOperationContext context, Size pulseLimit, ReplicationDocumentSender parent, ReplicationState replicationState, long next) : base(context, pulseLimit)
+            {
+                _parent = parent;
+                _replicationState = replicationState;
+                WasInterrupted = false;
+                Next = next;
+            }
+
+            public override void OnMoveNext(ReplicationBatchItem current)
+            {
+                _parent._parent.ForTestingPurposes?.OnDocumentSenderFetchNewItem?.Invoke();
+                _parent._parent.CancellationToken.ThrowIfCancellationRequested();
+                _parent.AssertNoLegacyReplicationViolation(current);
+
+                if (_replicationState.LastTransactionMarker != current.TransactionMarker)
+                {
+                    _replicationState.Item = current;
+                    long next = Next;
+                    if (_parent.CanContinueBatch(_replicationState, ref next) == false)
```

Comment:
Let's move `CanContinueBatch` into the `ReplicationBatchState` and unify it with the `_replicationState` ?

Reply:
Done